Proposal: First class secrets support in dev containers implementations (#224)
devcontainers/spec: proposals/secrets-support.md

## Goal

Support secrets as a first class feature in dev containers implementations.

This feature should consist of the ability to securely pass in the secrets, make them available for users to consume similar to remoteEnv and containerEnv, and be processed and handled securely at all times.
This functionality will allow consumers to be able to use secrets more predictably and securely with their dev containers implementation.

## Motivation

Today many consumers pass secret variables as `remoteEnv`, since there are no other mechanisms to pass them explicitly as secrets. The dev containers reference implementation do not (and need not) treat `remoteEnv` as secrets. Having explicit support for secrets will help eliminate security threats such as leaking secrets, as well as in improving customer confidence.

## Proposal

A [supporting tool](https://containers.dev/supporting#tools) (ie, for example the [dev container CLI reference implementation](https://github.com/devcontainers/cli)) should behave according to the following properties:

  1. Ability to pass secrets to commands
  2. Apply/use secrets similar to `remoteEnv`
  3. Securely handle secrets

### Passing secrets in

> Secrets are not part of dev containers specification and we do not expect users to store secrets inside devcontainer.json file. A conforming implemntation should provide a secure mechasism to pass secrets, such as a secrets file, Windows credential manager, Mac keychain, Azure keyvault etc. for example.

Reviewer: ```suggestion
Secrets are not part of dev containers specification and we do not expect users to store secrets inside `devcontainer.json` file. A conforming implementation should provide a secure mechanism to pass secrets, such as a secrets file, Windows credential manager, Mac keychain, Azure keyvault etc. for example.
```

# First class secrets support in devcontainers CLI

## What are secrets
Secrets are variables that hold sensitive values and needs to be handled securely at all times (API keys, passwords etc.). Users can change a secret's value at anytime, and the devcontainers CLI should support dynamically changing secrets without having to rebuild the container.

## Goal

Support Secrets as a first class feature in devcontainers CLI.

This feature should comprise of the ability to securely pass in the secrets in CLI commands, make them available for users to consume similar to remoteEnv and containerEnv, and be processed and handled securely at all times.
This functionality will allow consumers to be able to use secrets more predictably and securely with devcontainers CLI.

## Motivation

Today Codespaces pass secret variables as `remoteEnv` since there are no other mechanisms to pass them explicitly as secrets. The devcontainers CLI do not (and need not) treat `remoteEnv` as secrets. Having explicit support for secrets will help eliminating security threats such as leaking secrets, as well as in improving customer confidence.

## Proposal

> Following are the three main functional changes to the CLI to support this feature.

Reviewer 2: I know this is change is going to immediately be specific to the reference CLI implementation, but I think with slight rewording in a few key spots this could read more like a generalized specification.  Something like (IMHO):

```suggestion
A [supporting tool](https://containers.dev/supporting#tools) (ie, the [dev container CLI reference implementation](https://github.com/devcontainers/cli) should behave according to the following properties:
```

  1. Ability to pass secrets to commands
  2. Apply/use secrets similar to `remoteEnv`
  3. Securely handle secrets

### Passing secrets to commands

> Secrets are not part of devcontainers specification and we do not expect users to store secrets inside devcontainer.json file.

Reviewer: ```suggestion
Secrets are not part of dev containers specification and we do not expect users to store secrets inside devcontainer.json file.
```

We refer to dev containers as two words, can we update this?

> Secrets should be passed to the following devcontainers CLI commands.

Author: Sounds good. I will make the updates

> Secrets should be passed to the following devcontainers CLI commands.

Author: Made changes to make it a generic proposal

#### **Phase 1**
 - up
 - run-user-commands

> `up`, `run-user-commands` will use the secrets for usecases like lifecycle hooks (similar to remoteEnv). We do not plan secrets support for `build` command since we do not have a demostrated requirement, as well as to avoid potentially baking in the user secrets into the docker images.

Reviewer: ```suggestion
`up`, `run-user-commands` will use the secrets for usecases like lifecycle hooks (similar to remoteEnv). We do not plan secrets support for `build` command since we do not have a demonstrated requirement, as well as to avoid potentially baking in the user secrets into the docker images.
```

#### **Phase 2**
In phase 2 the support for secrets can be expanded to additional commands such as `set-up` and `exec`.

#### **Input format**

> Best way to pass secrets would be through a json/env file, where in the file path is supplied in a command line argument `--secrets-file`.

Author: Idea is that user can keep the secrets in a file (and store it anywhere on the machine, that they feel safe). It need not be inside devcontainer directory (and of course we don't expect them to check in this file.)
But to your point many of the secrets may be project specific and they could be stored along side other .devcontainer files. But I am not certain if we want to default to that.

## Future enhancements
Secondly, in the future we can expand to support secrets from additional sources such as Windows credential manager, Mac key chains, Azure keyvault etc. in addition to the secrets file support.
I think, to be flexible we might not want to limit the source to be the project dir.

> Best way to pass secrets would be through a json/env file, where in the file path is supplied in a command line argument `--secrets-file`.

Reviewer 3: Sill wondering how we are going to make it usable with the Dev Containers extension. Codespaces will get the secrets from GitHub's configuration for the repository. Will the Dev Containers extension have to introduce its own config file to associate secret files with devcontainer.json files?

Since each implementation can introduce its own solution for looking up secrets, the Dev Containers extension could look for a secrets file next to the devcontainer.json file without that location being part of the specification. But maybe this is an indication that we should add the default location to the OSS CLI? Otherwise the OSS CLI - when used locally - will not behave the same way the Dev Containers extension does.

> Best way to pass secrets would be through a json/env file, where in the file path is supplied in a command line argument `--secrets-file`.

Author: I have removed the CLI specific notes from this spec to keep it generic.

All CLI specific notes and recommendations has been moved to [an issue in CLI repo,](https://github.com/devcontainers/cli/issues/507) for consideration at the implementation time

Today the devcontaners CLI [commands accept](https://github.com/devcontainers/cli/blob/5c81479f0342947dd3e52a5984b9150e5feb8fd6/src/spec-node/devContainersSpecCLI.ts#L114) remote environment variables of the format name=value. However that format may not work as it is here, since in many cases the secret values may span multiple lines. One way to workaround that problem is by base64 encoding the values.

Example secrets file:
```
API_KEY=YWRzamhzZDZkZndkamZ3ZGU3ZWR3ZndlZGZkamVkd2Y3d2VkZndl
NUGET_CONFIG=PGNvbmZpZz4KICAgIDxhZGQga2V5PSJkZXBlbmRlbmN5VmVyc2lvbiIgdmFsdWU9IkhpZ2hlc3QiIC8+CiAgICA8YWRkIGtleT0iaHR0cF9wcm94eSIgdmFsdWU9Imh0dHA6Ly9jb21wYW55LXNxdWlkOjMxMjhAY29udG9zby5jb20iIC8+CjwvY29uZmlnPg==
PASSWORD=U2ltcGxlIFBhc3N3b3Jkcw==
```

Example secrets file in json format:
```json
[
  	{
		"name": "API_KEY",
		"value": "adsjhsd6dfwdjfwde7edwfwedfdjedwf7wedfwe"
	},
	{
		"name": "NUGET_CONFIG",
		"value": "<config>\n    <add key=\"dependencyVersion\" value=\"Highest\" />\n    <add key=\"http_proxy\" value=\"http://company-squid:3128@contoso.com\" />\n</config>"
	},
	{
		"name": "PASSWORD",
		"value": "Simple Passwords"
	}
]
```

> Further simplified json kv pair format (Preffered):

Reviewer: ```suggestion
Further simplified json kv pair format (Prefered):
```

```json
{
	"API_KEY": "adsjhsd6dfwdjfwde7edwfwedfdjedwf7wedfwe",
	"NUGET_CONFIG": "<config>\n    <add key=\"dependencyVersion\" value=\"Highest\" />\n    <add key=\"http_proxy\" value=\"http://company-squid:3128@contoso.com\" />\n</config>",
	"PASSWORD": "Simple Passwords"

> }

Author: Yes, for the CLI we will be adopting the simplified json format.
Do you think it still worth having the other two options documented for reference in this document?

```

### Using secrets in the CLI
Today Codespaces as a consumer is leveraging `remoteEnv` to pass secrets into the CLI. So it is a basic requirement that secrets should be used and behave the same way as `remoteEnv` do.

### Securely handle secrets
Secrets should be handled securely at all times.
- Secrets should not be passed as plain text in command line
- Do not be persist secrets on the container labels or long lived permanent caches

> - Redact secrets to prevent them being written in plain text into logs and console output.

Author: Potentially yes.
Fundamental requirement is that secret values must be masked before writing them to log files, so that sensitive values are not written in plan text to the log files and console

## Goal

Support secrets as a first class feature in dev containers implementations.

This feature should consist of the ability to securely pass in the secrets, make them available for users to consume similar to remoteEnv and containerEnv, and be processed and handled securely at all times.
This functionality will allow consumers to be able to use secrets more predictably and securely with their dev containers implementation.

## Motivation

Today many consumers pass secret variables as `remoteEnv`, since there are no other mechanisms to pass them explicitly as secrets. The dev containers reference implementation do not (and need not) treat `remoteEnv` as secrets. Having explicit support for secrets will help eliminate security threats such as leaking secrets, as well as in improving customer confidence.

## Proposal

A [supporting tool](https://containers.dev/supporting#tools) (ie, for example the [dev container CLI reference implementation](https://github.com/devcontainers/cli)) should behave according to the following properties:

  1. Ability to pass secrets to commands
  2. Apply/use secrets similar to `remoteEnv`
  3. Securely handle secrets

### Passing secrets in

> Secrets are not part of dev containers specification and we do not expect users to store secrets inside devcontainer.json file. It is recommended to pass the secrets in separate a file.

Reviewer 3: Should we call out that it is really up to the implementation where to get the secrets from and that the following are only examples? (E.g., Codespaces will use GitHub's secrets.)

I would then also only keep the short format we are going to implement first to keep the example in the proposal minimal. Maybe move the other formats to the notes further down if you want to keep them around.

Author: done
